### PR TITLE
Sign and verify support

### PR DESCRIPTION
**Pull request by Venemo**
This is currently a work-in-progress branch to implement sign and verify in the openssl crypto plugin. Opening pull request to get early feedback on the code.

NOTE: currently missing tests. Will add those soon.

### CODE REVIEW

File: plugins/opensslcryptoplugin/evp_p.h [hunk 1]
```diff
@@ -232,11 +232,57 @@ int osslevp_aes_decrypt_ciphertext(const unsigned char *init_vector,
...
     /* Clean up the decryption context */
-    EVP_CIPHER_CTX_cleanup(&decryption_context);
+    EVP_CIPHER_CTX_free(decryption_context);
     plaintext_length = update_length + final_length;
     return plaintext_length;
 }
 
+/*
+    Implements signing according to:
+    https://wiki.openssl.org/index.php/EVP_Signing_and_Verifying
```
Comment (chriadam) [inline plugins/opensslcryptoplugin/evp_p.h]: Add a comment to document the ownership semantics of the signature outparam, i.e. that the caller must free with OPENSSL_free()

File: plugins/opensslcryptoplugin/evp_p.h [hunk 2]
```diff
@@ -232,11 +232,57 @@ int osslevp_aes_decrypt_ciphertext(const unsigned char *init_vector,
...
+ */
+int osslevp_sign(const EVP_MD *digestFunc, EVP_PKEY *pkey, const void *bytes, size_t bytesCount, uint8_t **signature, size_t *signatureLength) {
+    EVP_MD_CTX *mdctx = EVP_MD_CTX_create();
+    int r;
```
Comment (chriadam) [inline plugins/opensslcryptoplugin/evp_p.h]: combine the declaration with the first use, i.e. int r = EVP_DigestSignInit(...)

File: plugins/opensslcryptoplugin/cryptoplugin_common.cpp [hunk 3]
```diff
@@ -407,15 +463,51 @@ CRYPTOPLUGINCOMMON_NAMESPACE::CRYPTOPLUGINCOMMON_CLASS::verify(
...
+                                        QLatin1String("Unsupported digest function chosen."));
+    }
+
+    // Verify the signature
+    r = osslevp_verify(evpDigestFunc, pkeyPtr, data.data(), data.length(), (const uint8_t*) signature.data(), (size_t) signature.length());
+    if (r == 1) {
+        // Verified successfully.
+        *verified = true;
+        return Sailfish::Crypto::Result(Sailfish::Crypto::Result::Succeeded);
+    }
```
Comment (chriadam) [inline plugins/opensslcryptoplugin/cryptoplugin_common.cpp]: Sailfish OS coding style https://sailfishos.org/wiki/Coding_Conventions requires the else on the same line as the closing-brace of the if block, i.e. } else if (...) {

File: plugins/opensslcryptoplugin/cryptoplugin_common.cpp [hunk 4]
```diff
@@ -407,15 +463,51 @@ CRYPTOPLUGINCOMMON_NAMESPACE::CRYPTOPLUGINCOMMON_CLASS::verify(
...
-                                    QLatin1String("TODO: verify"));
+    if (padding != Sailfish::Crypto::CryptoManager::SignaturePaddingNone) {
+        return Sailfish::Crypto::Result(Sailfish::Crypto::Result::UnsupportedOperation,
+                                        QLatin1String("TODO: signature padding other than None"));
+    }
+
+    QScopedPointer<BIO, LibCrypto_BIO_Deleter> bio(BIO_new(BIO_s_mem()));
+    QScopedPointer<EVP_PKEY, LibCrypto_EVP_PKEY_Deleter> pkey(EVP_PKEY_new());
+
+    int r;
```
Comment (chriadam) [inline plugins/opensslcryptoplugin/cryptoplugin_common.cpp]: combine declaration of r with its first use, e.g. int r = BIO_write(...);

File: plugins/opensslcryptoplugin/cryptoplugin_common.cpp [hunk 5]
```diff
@@ -388,14 +406,52 @@ CRYPTOPLUGINCOMMON_NAMESPACE::CRYPTOPLUGINCOMMON_CLASS::sign(
...
-                                    QLatin1String("TODO: sign"));
+    if (padding != Sailfish::Crypto::CryptoManager::SignaturePaddingNone) {
+        return Sailfish::Crypto::Result(Sailfish::Crypto::Result::UnsupportedOperation,
+                                        QLatin1String("TODO: signature padding other than None"));
+    }
+
+    QScopedPointer<BIO, LibCrypto_BIO_Deleter> bio(BIO_new(BIO_s_mem()));
+    QScopedPointer<EVP_PKEY, LibCrypto_EVP_PKEY_Deleter> pkey(EVP_PKEY_new());
+
+    int r;
```
Comment (chriadam) [inline plugins/opensslcryptoplugin/cryptoplugin_common.cpp]: combine declaration of r with its first usage, e.g. int r = BIO_write(..)

File: plugins/opensslcryptoplugin/cryptoplugin_common.cpp [hunk 6]
```diff
@@ -101,6 +108,17 @@ namespace {
         }
         return 0; // no cipher sessions available.
     }
+
+    static const EVP_MD *getEvpDigestFunction(Sailfish::Crypto::CryptoManager::DigestFunction digestFunction) {
```
Comment (chriadam) [inline plugins/opensslcryptoplugin/cryptoplugin_common.cpp]: in anonymous namespace so doesn't need static qualifier I guess

**Comment by chriadam** [comment]
Once Bea has finished with https://github.com/sailfishos/sailfish-secrets/pull/31/files you should rebase (her commit reduces code-duplication (evp.h) which you noticed).

**Comment by chriadam** [comment]
Mostly LGTM - thanks very much.  Obviously needs unit tests, but mostly looking great.

**Comment by chriadam** [comment]
I wonder whether I need to extend the function signatures to also include optional MessageAuthenticationCode and Algorithm parameters, so that we can use HMAC (https://www.openssl.org/docs/man1.0.2/crypto/hmac.html) or CMAC (http://rubenlaguna.com/post/2015-02-05-compute-aes-cmac-using-openssl-slash-libcrypto/) if required.

File: plugins/opensslcryptoplugin/evp_p.h (hunk 1 above)
Comment (Venemo) [reply plugins/opensslcryptoplugin/evp_p.h]: Ok!

File: plugins/opensslcryptoplugin/evp_p.h (hunk 2 above)
Comment (Venemo) [reply plugins/opensslcryptoplugin/evp_p.h]: I usually don't, when the variable is then used for different stuff repeatedly. But I'll change it.

File: plugins/opensslcryptoplugin/cryptoplugin_common.cpp (hunk 6 above)
Comment (Venemo) [reply plugins/opensslcryptoplugin/cryptoplugin_common.cpp]: No, it indeed doesn't.

**Comment by Venemo** [comment]
@chriadam For the HMAC and CMAC, I would suggest to deal with those separately after we managed to get the current stuff working.

File: plugins/opensslcryptoplugin/cryptoplugin_common.cpp (hunk 3 above)
Comment (Venemo) [reply plugins/opensslcryptoplugin/cryptoplugin_common.cpp]: Sorry, I'm not used to the TIE-fighter else, but will change it!

**Comment by Venemo** [comment]
Amended my previous commit.

**Comment by Venemo** [comment]
After the plugin load fix, I am able to run it on my desktop like this:

```sh
LD_LIBRARY_PATH=../lib/Secrets:../lib/Crypto QT_PLUGIN_PATH=../tests/plugins/testopensslplugin:../tests/plugins/testopensslcryptoplugin:../tests/plugins/testsqlcipherplugin:../tests/plugins/testsqliteplugin ./sailfishsecretsd --test
```

And execute the tests like this:

```sh
LD_LIBRARY_PATH=/home/Timur/dev/Qt/5.6.3/gcc_64/lib:../../../lib/Secrets:../../../lib/Crypto ./tst_cryptorequests
```

**Comment by chriadam** [comment]
How much overhead does the new plugin loading semantic imply?  That is, on Sailfish OS devices, how many more libraries are speculatively loaded as potential Secrets/Crypto plugins?

**Comment by Venemo** [comment]
Closing this PR, will open a new one.

**Pull request closed without merging**